Fix autograd.Function + NJT when an output grad is None

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #136792
* __->__ #136875

For `autograd.Function`, the engine will try to allocate correctly-shaped zeros for `None` grads (i.e. in the case where the output isn't used downstream). It determines the shape of these zeros from the `VariableInfo` entry, which is derived from the forward output shape. For the NJT forward output case, the size info stored will contain a nested int, and calling `zeros()` with this size throws:
```
RuntimeError: .../build/aten/src/ATen/RegisterCPU.cpp:5260: SymIntArrayRef expected to contain only concrete integers
```

This PR fixes this by storing the full tensor in the `VariableInfo` for the nested case and calling `zeros_like()` to allocate correctly-shaped zeros. This is pretty inefficient; ideally we would want to save just the NJT shape and be able to construct zeros from it, but this requires factory function support for nested ints (WIP). So this is a short-term fix until we have that.

jbschlosser committed:
Fix autograd.Function + NJT when an output grad is None

[ghstack-poisoned]

soulitzer commented:
Maybe we could do this under a flag, since I'd consider it a relatively rare case that can be worked around by the user in most situations.

jbschlosser commented:
> Maybe we could do this under a flag, since I'd consider it a relatively rare case that can be worked around by the user in most situations.

Which part would be under the flag? Would the maintainer of e.g. `FlexAttentionAutogradOp` be the one responsible for setting it? Seems a bit outside of the scope there, as I'm sure they just want things to work wrt gradients.

Note that I ran into this in practice during integration with FlexAttention (`logsumexp` is optionally used downstream, with the default being unused). This change is only done for nested tensors by default, so I don't expect too much of an impact on dense tensors.

soulitzer commented:
I think we should not save when there's only a single output. And then we can offer a flag that whoever is calling the autograd Function can flip to remove the extra saving, promising that they will not.

soulitzer commented:
I think its an unfortunate situation either way, but...probably fine and we shouldn't stress too much since we'll have factory functions hopefully soon :P

Cons:
The "opt-in" approach does not work out of the box is but easy to debug if goes something goes wrong. The "opt-out" approach is harder to debug because all you know is that you are now using more memory. One particular example I'm thinking of is compile, where now if you switch from eager to compile your model may use more memory and oom.
 
Pros:
The flip side is that NJT is still a prototype and in the stage of rapidly acquiring new users - the opt-in approach makes lives easier for people who weren't using NJT before and the claim is that for existing users the increased memory use is not that visible.

One I would say is that we can further decrease the blast radius of this PR by limiting to only save when there are at least two differentiable outputs, and one of them is the NJT. We could add a flag to disable this for the cases where the user is certain that they won't run into this, but I'm also not sure where to put that.

jbschlosser commented:
> the opt-in approach makes lives easier for people who weren't using NJT before

sorry I'm confused about this - why would an opt-in approach serve people who weren't using NJT before? this PR doesn't affect them since it only makes changes conditional on nested tensors.

> The "opt-out" approach is harder to debug because all you know is that you are now using more memory.

By "more memory" here, do you mean for the case where an autograd.Function utilizes NJT but doesn't hit the "output grad is None" case?

> One I would say is that we can further decrease the blast radius of this PR by limiting to only save when there are at least two differentiable outputs, and one of them is the NJT. 

sure, I think this should cover the prior case pretty well

jbschlosser committed:
Update on "Fix autograd.Function + NJT when an output grad is None"


For `autograd.Function`, the engine will try to allocate correctly-shaped zeros for `None` grads (i.e. in the case where the output isn't used downstream). It determines the shape of these zeros from the `VariableInfo` entry, which is derived from the forward output shape. For the NJT forward output case, the size info stored will contain a nested int, and calling `zeros()` with this size throws:
```
RuntimeError: .../build/aten/src/ATen/RegisterCPU.cpp:5260: SymIntArrayRef expected to contain only concrete integers
```

This PR fixes this by storing the full tensor in the `VariableInfo` for the nested case and calling `zeros_like()` to allocate correctly-shaped zeros. This is pretty inefficient; ideally we would want to save just the NJT shape and be able to construct zeros from it, but this requires factory function support for nested ints (WIP). So this is a short-term fix until we have that.

[ghstack-poisoned]

soulitzer commented:
> sorry I'm confused about this - why would an opt-in approach serve people who weren't using NJT before? this PR doesn't affect them since it only makes changes conditional on nested tensors.

Ah typo, I meant "opt out" lol

soulitzer reviewed (approved):
OK!

jbschlosser commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -733,8 +733,14 @@ static void _wrap_outputs(
         // NOLINTNEXTLINE(bugprone-unchecked-optional-access)
-        self->output_info.emplace_back(*wrapped_outputs[i]);
+        bool use_zeros_like =
+            num_outputs > 1 && wrapped_outputs[i]->is_nested();
```
ah true I forgot the differentiable type check - will do that

jbschlosser committed:
Update on "Fix autograd.Function + NJT when an output grad is None"


For `autograd.Function`, the engine will try to allocate correctly-shaped zeros for `None` grads (i.e. in the case where the output isn't used downstream). It determines the shape of these zeros from the `VariableInfo` entry, which is derived from the forward output shape. For the NJT forward output case, the size info stored will contain a nested int, and calling `zeros()` with this size throws:
```
RuntimeError: .../build/aten/src/ATen/RegisterCPU.cpp:5260: SymIntArrayRef expected to contain only concrete integers
```

This PR fixes this by storing the full tensor in the `VariableInfo` for the nested case and calling `zeros_like()` to allocate correctly-shaped zeros. This is pretty inefficient; ideally we would want to save just the NJT shape and be able to construct zeros from it, but this requires factory function support for nested ints (WIP). So this is a short-term fix until we have that.

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser committed:
Update on "Fix autograd.Function + NJT when an output grad is None"


For `autograd.Function`, the engine will try to allocate correctly-shaped zeros for `None` grads (i.e. in the case where the output isn't used downstream). It determines the shape of these zeros from the `VariableInfo` entry, which is derived from the forward output shape. For the NJT forward output case, the size info stored will contain a nested int, and calling `zeros()` with this size throws:
```
RuntimeError: .../build/aten/src/ATen/RegisterCPU.cpp:5260: SymIntArrayRef expected to contain only concrete integers
```

This PR fixes this by storing the full tensor in the `VariableInfo` for the nested case and calling `zeros_like()` to allocate correctly-shaped zeros. This is pretty inefficient; ideally we would want to save just the NJT shape and be able to construct zeros from it, but this requires factory function support for nested ints (WIP). So this is a short-term fix until we have that.

[ghstack-poisoned]

huydhn commented on torch/csrc/autograd/variable_info.cpp:
```diff
@@ -2,27 +2,31 @@
       requires_grad(var.requires_grad()),
-      is_empty(false) {}
+      is_empty(false),
+      the_var(use_zeros_like ? std::optional<Variable>(var) : std::nullopt) {}
```
@jbschlosser I think when `use_zeros_like` is set to True here, it causes a memory leak in `test_sdpa_backwards` trunk, i.e. https://github.com/pytorch/pytorch/actions/runs/11299116618/job/31430838672#step:22:1500

```
PYTORCH_TEST_CUDA_MEM_LEAK_CHECK=1 python test/test_nestedtensor.py TestNestedTensorSubclassCUDA.test_sdpa_backwards_cuda_float16

/home/huydo/local/conda/py3.9/lib/python3.9/site-packages/torch/backends/cudnn/__init__.py:108: UserWarning: PyTorch was compiled without cuDNN/MIOpen support. To use cuDNN/MIOpen, rebuild PyTorch making sure the library is visible to the build system.
  warnings.warn(
/home/huydo/local/conda/py3.9/lib/python3.9/site-packages/torch/backends/cudnn/__init__.py:108: UserWarning: PyTorch was compiled without cuDNN/MIOpen support. To use cuDNN/MIOpen, rebuild PyTorch making sure the library is visible to the build system.
  warnings.warn(
W1011 20:42:23.899092 380389 /data/users/huydo/conda/py3.9/lib/python3.9/site-packages/torch/_dynamo/variables/tensor.py:832] [0/0] Graph break from `Tensor.item()`, consider setting:
W1011 20:42:23.899092 380389 /data/users/huydo/conda/py3.9/lib/python3.9/site-packages/torch/_dynamo/variables/tensor.py:832] [0/0]     torch._dynamo.config.capture_scalar_outputs = True
W1011 20:42:23.899092 380389 /data/users/huydo/conda/py3.9/lib/python3.9/site-packages/torch/_dynamo/variables/tensor.py:832] [0/0] or:
W1011 20:42:23.899092 380389 /data/users/huydo/conda/py3.9/lib/python3.9/site-packages/torch/_dynamo/variables/tensor.py:832] [0/0]     env TORCHDYNAMO_CAPTURE_SCALAR_OUTPUTS=1
W1011 20:42:23.899092 380389 /data/users/huydo/conda/py3.9/lib/python3.9/site-packages/torch/_dynamo/variables/tensor.py:832] [0/0] to include these operations in the captured graph.
W1011 20:42:23.899092 380389 /data/users/huydo/conda/py3.9/lib/python3.9/site-packages/torch/_dynamo/variables/tensor.py:832] [0/0]
W1011 20:42:23.899092 380389 /data/users/huydo/conda/py3.9/lib/python3.9/site-packages/torch/_dynamo/variables/tensor.py:832] [0/0] Graph break: from user code at:
W1011 20:42:23.899092 380389 /data/users/huydo/conda/py3.9/lib/python3.9/site-packages/torch/_dynamo/variables/tensor.py:832] [0/0]   File "/data/users/huydo/github/pytorch/test/test_nestedtensor.py", line 6719, in f
W1011 20:42:23.899092 380389 /data/users/huydo/conda/py3.9/lib/python3.9/site-packages/torch/_dynamo/variables/tensor.py:832] [0/0]     torch.tensor(1).item()
W1011 20:42:23.899092 380389 /data/users/huydo/conda/py3.9/lib/python3.9/site-packages/torch/_dynamo/variables/tensor.py:832] [0/0]
W1011 20:42:23.899092 380389 /data/users/huydo/conda/py3.9/lib/python3.9/site-packages/torch/_dynamo/variables/tensor.py:832] [0/0]
/home/huydo/local/conda/py3.9/lib/python3.9/site-packages/torch/testing/_internal/common_utils.py:2366: UserWarning: CUDA caching allocator reports a memory leak not verified by the driver API in __main__.TestNestedTensorSubclassCUDA.test_sdpa_backwards_cuda_float16! Caching allocator allocated memory was 0 and is now reported as 28160 on device 0. CUDA driver allocated memory was 21747138560 and is now 21745041408.
  warnings.warn(msg)
.
----------------------------------------------------------------------
Ran 1 test in 72.632s

OK
```

Could you help take a look to confirm?

jbschlosser commented:
@pytorchbot revert -m "Caused memory leak" -c nosignal

jbschlosser committed:
Update on "Fix autograd.Function + NJT when an output grad is None"


For `autograd.Function`, the engine will try to allocate correctly-shaped zeros for `None` grads (i.e. in the case where the output isn't used downstream). It determines the shape of these zeros from the `VariableInfo` entry, which is derived from the forward output shape. For the NJT forward output case, the size info stored will contain a nested int, and calling `zeros()` with this size throws:
```
RuntimeError: .../build/aten/src/ATen/RegisterCPU.cpp:5260: SymIntArrayRef expected to contain only concrete integers
```

This PR fixes this by storing the full tensor in the `VariableInfo` for the nested case and calling `zeros_like()` to allocate correctly-shaped zeros. This is pretty inefficient; ideally we would want to save just the NJT shape and be able to construct zeros from it, but this requires factory function support for nested ints (WIP). So this is a short-term fix until we have that.

[ghstack-poisoned]

jbschlosser commented:
Fix is in to detach properly @soulitzer . I verified that it fixes the broken test.

soulitzer reviewed (approved):
Thanks!

jbschlosser commented:
@pytorchbot merge
